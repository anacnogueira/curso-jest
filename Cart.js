import find from 'lodash/find';
import remove from 'lodash/remove';
import Dinero from 'dinero.js';

const calculatePercentageDiscount = (amount, { condition, quantity }) => {
  if (condition?.percentage && quantity > condition.minimum) {
    return amount.percentage(condition.percentage);
  }

  return Money({ amount: 0 });
};

const calculateQuantityDiscount = (amount, { condition, quantity }) => {
  const isEven = quantity % 2 === 0;

  if (condition?.quantity && quantity > condition.quantity) {
    return amount.percentage(isEven ? 50 : 40);
  }

  return Money({ amount: 0 });
};

const calculateDiscount = (amount, quantity, condition) => {
  const list = Array.isArray(condition) ? condition : [condition];

  const [higherDiscount] = list
    .map(cond => {
      if (cond.percentage) {
        return calculatePercentageDiscount(amount, {
          condition: cond,
          quantity,
        }).getAmount();
      } else if (cond.quantity) {
        return calculateQuantityDiscount(amount, {
          condition: cond,
          quantity,
        }).getAmount();
      }
    })
    .sort((a, b) => b - a);

  return Money({ amount: higherDiscount });
};

const Money = Dinero;
Money.defaultCurrency = 'BRL';
Money.defaultPrecision = 2;

export default class Cart {
  items = [];

  add(item) {
    const itemToFind = { product: item.product };

    if (find(this.items, itemToFind)) {
      remove(this.items, itemToFind);
    }

    this.items.push(item);
  }

  remove(product) {
    remove(this.items, { product });
  }

  getTotal() {
    return this.items.reduce((acc, { quantity, product, condition }) => {
      const amount = Money({ amount: quantity * product.price });
      let discount = Money({ amount: 0 });

      if (condition) {
        discount = calculateDiscount(amount, quantity, condition);
      }

      return acc.add(amount).subtract(discount);
    }, Money({ amount: 0 }));
  }

  summary() {
    const total = this.getTotal();
    const formatted = total.toFormat('$0,0.00');
    const items = this.items;

    return {
      total,
      formatted,
      items,
    };
  }

  checkout() {
    const { total, items } = this.summary();

    this.items = [];

    return {
      total: total.getAmount(),
      items,
    };
  }
}
