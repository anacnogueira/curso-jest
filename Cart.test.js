import Cart from './Cart';

describe('Cart', () => {
  test('should return 0 when getTotal() is executed in a newly created instance ', () => {
    const cart = new Cart();
    expect(cart.getTotal()).toEqual(0);
  });
});
