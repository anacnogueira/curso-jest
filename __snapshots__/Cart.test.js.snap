// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the list of items 1`] = `
{
  "items": [
    {
      "product": {
        "price": 35388,
        "title": "Adidas Running Shoes Pink - Women",
      },
      "quantity": 2,
    },
    {
      "product": {
        "price": 41872,
        "title": "Adidas Running Shoes Pink - men",
      },
      "quantity": 3,
    },
  ],
  "total": 196392,
}
`;

exports[`Cart checkout() should return an object with the total and the list of items when summary is called 1`] = `
{
  "formatted": "R$1,963.92",
  "items": [
    {
      "product": {
        "price": 35388,
        "title": "Adidas Running Shoes Pink - Women",
      },
      "quantity": 2,
    },
    {
      "product": {
        "price": 41872,
        "title": "Adidas Running Shoes Pink - men",
      },
      "quantity": 3,
    },
  ],
  "total": {
    "amount": 196392,
    "currency": "BRL",
    "precision": 2,
  },
}
`;
